fabric_generator/fabric_gen.py

```diff
@@ -15,6210 +15,2273 @@
-    return temp_list
-
-def GetTileComponentPort_Verilog (tile_description, port_direction, port_prefix=''):
-    ports = []
-    for line in tile_description:
-        if line[0] == port_direction:
-            if line[source_name] != 'NULL':
-                if port_prefix:
-                    ports.append(port_prefix+'_'+line[source_name])
-                else:
-                    ports.append(line[source_name])
-    for line in tile_description:
-        if line[0] == Opposite_Directions[port_direction]:
-            if line[destination_name] != 'NULL':
-                if port_prefix:
-                    ports.append(port_prefix+'_'+line[destination_name])
-                else:
-                    ports.append(line[destination_name])
-    return ports
-
-def GenerateVerilog_Conf_Instantiation ( file, counter, close=True ):
-    print('\t//GLOBAL all primitive pins for configuration (not further parsed)', file=file)
-    print('\t.MODE(Mode),', file=file)
-    print('\t.CONFin(conf_data['+str(counter)+']),', file=file)
-    print('\t.CONFout(conf_data['+str(counter+1)+']),', file=file)
-    if close==True:
-        print('\t.CLK(CLK)', file=file)
-        print('\t);\n', file=file)
-    else:
-        print('\t.CLK(CLK),', file=file)
-    return
-
-def GetVerilogDeclarationForFile(VHDL_file_name):
-    ConfigPortUsed = 0 # 1 means is used
-    VHDLfile = [line.rstrip('\n') for line in open(VHDL_file_name)]
-    templist = []
-    # for item in VHDLfile:
-        # print(item)
-    for line in VHDLfile:
-        # NumberOfConfigBits:0 means no configuration port
-        if re.search('NumberOfConfigBits', line, flags=re.IGNORECASE):
-            # NumberOfConfigBits appears, so we may have a config port
-            ConfigPortUsed = 1
-            # but only if the following is not true
-            if re.search('NumberOfConfigBits:0', line, flags=re.IGNORECASE):
-                ConfigPortUsed = 0
-    #print('', file=file)
-    return ConfigPortUsed
-
-#CAD methods from summer vacation project 2020 by Bea
-sDelay = "8"
-GNDRE = re.compile("GND(\d*)")
-VCCRE = re.compile("VCC(\d*)")
-VDDRE = re.compile("VDD(\d*)")
-BracketAddingRE = re.compile(r"^(\S+?)(\d+)$")
-letters = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W"] #For LUT labelling
-
-#This class represents individual tiles in the architecture
-class Tile:
-    tileType = ""
-    bels = []
-    belsWithIO = [] #Currently the plan is to deprecate bels and replace it with this. However, this would require nextpnr model generation changes, so I won't do that until the VPR foundations are established
-    #Format for belsWithIO is [bel name, prefix, inputs, outputs, whether it has a clock input]
-    #Format for bels is [bel name, prefix, ports, whether it has a clock input]
-    wires = [] 
-    atomicWires = [] #For storing single wires (to handle cascading and termination)
-    pips = []
-   
```

Comment:
I think `len(connections[portName]).bit_length()` should work here without going through FP